modules/renter/siafile/siafile.go

```diff
@@ -0,0 +1,282 @@
+package siafile
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"os"
+	"sync"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/fastrand"
+
+	"github.com/NebulousLabs/Sia/crypto"
+)
+
+type (
+	// SiaFile is the disk format for files uploaded to the Sia network.  It
+	// contains all the necessary information to recover a file from its hosts and
+	// allows for easy constant-time updates of the file without having to read or
+	// write the whole file.
+	SiaFile struct {
+		// staticMetadata is the mostly static staticMetadata of a SiaFile. The reserved
+		// size of the staticMetadata on disk should always be a multiple of 4kib.
+		// The staticMetadata is also the only part of the file that is JSON encoded
+		// and can therefore be easily extended.
+		staticMetadata Metadata
+
+		// pubKeyTable stores the public keys of the hosts this file's pieces are uploaded to.
+		// Since multiple pieces from different chunks might be uploaded to the same host, this
+		// allows us to deduplicate the rather large public keys.
+		pubKeyTable []types.SiaPublicKey
+
+		// staticChunks are the staticChunks the file was split into.
+		staticChunks []Chunk
+
+		// utility fields. These are not persisted.
+		deleted   bool
+		mu        sync.RWMutex
+		staticUID string
+	}
+
+	// Chunk represents a single chunk of a file on disk
+	Chunk struct {
+		// erasure code settings.
+		//
+		// staticErasureCodeType specifies the algorithm used for erasure coding
+		// chunks. Available types are:
+		//   0 - Invalid / Missing Code
+		//   1 - Reed Solomon Code
+		//
+		// erasureCodeParams specifies possible parameters for a certain
+		// staticErasureCodeType. Currently params will be parsed as follows:
+		//   Reed Solomon Code - 4 bytes dataPieces / 4 bytes parityPieces
+		//
+		staticErasureCodeType   [4]byte
+		staticErasureCodeParams [8]byte
+		staticErasureCode       modules.ErasureCoder
+
+		// extensionInfo is some reserved space for each chunk that allows us
+		// to indicate if a chunk is special.
+		extensionInfo [16]byte
+
+		// pieces are the pieces of the file the chunk consists of.
+		pieces [][]Piece
+	}
+
+	// Piece represents a single piece of a chunk on disk
+	Piece struct {
+		KeyNonce   [4]byte            // nonce used for encrypting the piece
+		HostPubKey types.SiaPublicKey // public key of the host
```

Comment:
Can we change that in the PR that changes the persistence? Right now it would only make the code in `compat.go` more complicated since the table doesn't have a fixed ordering.